File: src/segment-queue.ts
export interface SegmentQueueOptions {
  synthesize: (text: string) => Promise<void>;
  signal: AbortSignal;
  onError?: (err: Error) => void;
}

/**
 * Async FIFO of text segments with a single serial worker that calls the
 * provided `synthesize` callback for each segment in order. Used by the
 * pipeline to keep TTS playback strictly ordered while segments are being
 * produced upstream by the idle-segmenter.
 */
export class SegmentQueue {
  private items: string[] = [];
  private closed = false;
  private wakeup: (() => void) | null = null;
  private workerPromise: Promise<void>;

  constructor(private opts: SegmentQueueOptions) {
    this.workerPromise = this.run();
  }

  push(segment: string): void {
    if (this.closed || segment.length === 0) return;
    this.items.push(segment);
    this.notify();
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.notify();
  }

  /** Resolves when the worker has drained the queue (or aborted). */
  done(): Promise<void> {
    return this.workerPromise;
  }

  private notify(): void {
    if (this.wakeup) {
      const w = this.wakeup;
      this.wakeup = null;
      w();
    }
  }

  private async run(): Promise<void> {
    while (true) {
      if (this.opts.signal.aborted) return;

      if (this.items.length === 0) {
        if (this.closed) return;
        await new Promise<void>((resolve) => {
          this.wakeup = resolve;
        });
        continue;
      }

      const segment = this.items.shift()!;
      const pending = this.items.length;
      console.log(`[segment-queue] synthesizing (${segment.length} chars, ${pending} queued): "${segment.substring(0, 80)}${segment.length > 80 ? "..." : ""}"`);
      try {
        await this.opts.synthesize(segment);
        console.log(`[segment-queue] done (${segment.length} chars)`);
      } catch (err) {
        if (this.opts.onError) {
          this.opts.onError(err as Error);
        } else {
          console.error("[segment-queue] synth error:", (err as Error).message);
        }
      }
    }
  }
}

File: src/vauxr/segment_queue.py
"""Async FIFO of text segments with a single serial synthesizer worker.

Port of `src/segment-queue.ts`. Used by the pipeline to keep TTS playback
strictly ordered while segments are produced upstream by the
:mod:`vauxr.idle_segmenter`.
"""

from __future__ import annotations

import asyncio
import logging
from collections.abc import Awaitable, Callable

log = logging.getLogger("vauxr.segment_queue")


class SegmentQueue:
    def __init__(
        self,
        *,
        synthesize: Callable[[str], Awaitable[None]],
        abort_event: asyncio.Event,
        on_error: Callable[[BaseException], None] | None = None,
    ) -> None:
        self._synthesize = synthesize
        self._abort = abort_event
        self._on_error = on_error
        self._items: list[str] = []
        self._closed = False
        self._wakeup: asyncio.Event = asyncio.Event()
        self._worker = asyncio.create_task(self._run())

    def push(self, segment: str) -> None:
        if self._closed or not segment:
            return
        self._items.append(segment)
        self._wakeup.set()

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._wakeup.set()

    async def done(self) -> None:
        await self._worker

    async def _run(self) -> None:
        while True:
            if self._abort.is_set():
                return
            if not self._items:
                if self._closed:
                    return
                self._wakeup.clear()
                # Wait for either a new item, close(), or abort.
                abort_wait = asyncio.create_task(self._abort.wait())
                wake_wait = asyncio.create_task(self._wakeup.wait())
                try:
                    done, pending = await asyncio.wait(
                        {abort_wait, wake_wait}, return_when=asyncio.FIRST_COMPLETED
                    )
                finally:
                    for t in (abort_wait, wake_wait):
                        if not t.done():
                            t.cancel()
                if self._abort.is_set():
                    return
                continue

            segment = self._items.pop(0)
            pending = len(self._items)
            log.info(
                "synthesizing (%d chars, %d queued): %r",
                len(segment),
                pending,
                segment[:80] + ("…" if len(segment) > 80 else ""),
            )
            try:
                await self._synthesize(segment)
                log.info("done (%d chars)", len(segment))
            except Exception as err:  # noqa: BLE001
                if self._on_error is not None:
                    self._on_error(err)
                else:
                    log.error("synth error: %s", err)
